RULES/core/context.go

```diff
@@ -93,13 +119,41 @@ func newContext(vars map[string]interface{}) *context {
 	return ctx
 }
 
+func (ctx *context) Build(target buildInterface) BuildOutput {
+	flagStrings := []string{}
+	for name, value := range ctx.flags {
+		flagStrings = append(flagStrings, fmt.Sprintf("flag:%s$%s", name, value))
+	}
+	for param, option := range ctx.buildOptions {
+		flagStrings = append(flagStrings, fmt.Sprintf("buildOption:%s$%s", param.Name, option.(buildOptionName).Name()))
+	}
+	sort.Strings(flagStrings)
+	ctx.hash = fmt.Sprintf("%08X", crc32.ChecksumIEEE([]byte(strings.Join(flagStrings, "#"))))
```

Comment:
A potential issue is that since the whole ctx is hashed when we build a target, and not just the BuildOptions that are relevant, if the same target is built with the same relevant BuildOption, but some irrelevant ones have changed, it will be built twice unnecessarily. This is not optimal, and might lead to something like duplicate symbols at link-time. I'm not 100% sure about this, I might be missing something.